charmcraft/config.py

```diff
@@ -0,0 +1,152 @@
+            "Bad charmcraft.yaml content; field '{}' (if present) must value 'charm' or 'bundle': "
+            "got {!r}.".format(key, value))
+
+
+@attr.s(kw_only=True, frozen=True)
+class _CharmhubConfig:
+    section = 'charmhub'
+
+    api_url = attr.ib(
+        type=str, default='https://api.staging.charmhub.io',
+        validator=[_check_url, _check_type])
+    storage_url = attr.ib(
+        type=str, default='https://storage.staging.snapcraftcontent.com',
+        validator=[_check_url, _check_type])
+
+    @classmethod
+    def from_dict(cls, source=None):
+        """Build from a raw dict."""
+        if source is None:
+            return cls()
+
+        if not isinstance(source, dict):
+            raise CommandError(
+                "Bad charmcraft.yaml content; the 'charmhub' field must be a dict: got {!r}."
+                .format(type(source).__name__))
+        return cls(**source)
+
+
+class _BasicPrime(tuple):
+    """Hold the list of files to include, specified under parts/bundle/prime configs.
+
+    This is a intermediate structure until we have the full Lifecycle in place.
+    """
+    @classmethod
+    def from_dict(cls, parts):
+        """Build from a dicts sequence."""
+        if not isinstance(parts, dict):
+            raise CommandError(
+                "Bad charmcraft.yaml content; the 'parts' field must be a dict: got {!r}."
+                .format(type(parts).__name__))
+
+        bundle = parts.get('bundle', {})
+        if not isinstance(bundle, dict):
+            raise CommandError(
+                "Bad charmcraft.yaml content; the 'parts.bundle' field must be a dict: got {!r}."
+                .format(type(bundle).__name__))
+
+        prime = bundle.get('prime', [])
+        if not isinstance(prime, list):
+            raise CommandError(
+                "Bad charmcraft.yaml content; the 'parts.bundle.prime' field must be a list: "
+                "got {!r}.".format(type(prime).__name__))
+
+        # validate that all are relative
+        for spec in prime:
+            # check if it's an absolute path using POSIX's '/' (not os.path.sep, as the charm's
+            # config is independent of the platform where charmcraft is running)
+            if spec[0] == '/':
+                raise CommandError(
+                    "Bad charmcraft.yaml content; the paths specifications in "
+                    "'parts.bundle.prime' must be relative: found {!r}.".format(spec))
+        return cls(prime)
+
+
+@attr.s(kw_only=True, frozen=True)
+class Config:
+    charmhub = attr.ib(default={}, converter=_CharmhubConfig.from_dict)
+    parts = attr.ib(default={}, converter=_BasicPrime.from_dict)
+    type = attr.ib(default=None, validator=[_check_typefield])
+
+    # this value is provided by the code itself, not the user, as convenience for the
+    # rest of the code
+    project_dirpath = attr.ib(default=None)
+
+    @classmethod
+    def from_file(cls, project_directory):
```

Comment:
deconflict "from_file" vs directory argument. perhaps use the chamcraft file path as arg?

Reply:
Will decouple it even more. I will put a "load" function at module level.